packages/kernel-utils/src/object-disjoint-union.ts

```diff
@@ -0,0 +1,34 @@
+/**
+ * Create a new object that is the disjoint union of the given objects.
+ *
+ * @param objects - The objects to union.
+ * @returns The disjoint union of the given objects.
+ * @throws If a key is found in multiple objects.
+ */
+export const objectDisjointUnion = (
+  ...objects: Record<PropertyKey, unknown>[]
+): Record<PropertyKey, unknown> => {
+  const keys = new Map<PropertyKey, number>();
+  const out: Record<PropertyKey, unknown> = Object.create(null);
+  let collidingIndex = 0;
+  for (const obj of objects) {
+    for (const key of Reflect.ownKeys(obj)) {
+      if (keys.has(key)) {
+        const originalIndex = keys.get(key);
```

Comment:
```suggestion
   objects.forEach((obj, originalIndex) => {
      if (keys.has(key)) {
```

I guess this is simpler?